### ISSUE TITLE
Error "Sequence contains no elements..."

### ISSUE THREAD

**Issue by SujataZ** [sethjuarez/numl#66]
Hi Seth,

I get this error at random times with the same data. It comes either at the learning step in below code -  
 generator = new DecisionTreeGenerator(5) { Descriptor = descriptor, Hint = 0 };
learningModel = Learner.Learn(data, 0.8, 5, generator);

Or
While deserilizing the model json like 

JsonReader.ReadJson<T>(jsonData);

This error really comes randomly & goes away with the same data after a few retries.

Can you please help spotting it?

**Comment by sethjuarez** [comment]
Weird, do you have a stack trace? It could either be features in the descriptor or elements in the training set.

**Comment by SujataZ** [comment]
Don't have the stack trace currently, but for learning it comes from "ToVector" method & for json reading, comes from the "Read" method itself...
Will post the stack trace as soon as I get the error again

**Comment by SujataZ** [comment]
Seems that you have upgraded the project to VS2017, unable to reference the source code for debugging issues with my code now... 

Do you have the older VS project for version 0.9.15-beta?

**Comment by SujataZ** [comment]
Here is the stacktrace for "Sequence contains no elements" error while calling the "Learn" method

 at System.Linq.Enumerable.First[TSource](IEnumerable`1 source)
   at numl.Model.Descriptor.At(Int32 i)
   at numl.Supervised.DecisionTree.DecisionTreeGenerator.GetBestSplit(Matrix x, Vector y, List`1 used)
   at numl.Supervised.DecisionTree.DecisionTreeGenerator.BuildTree(Matrix x, Vector y, Int32 depth, List`1 used, Tree tree)
   at numl.Supervised.DecisionTree.DecisionTreeGenerator.Generate(Matrix X, Vector y)
   at numl.Learner.GenerateModel(IGenerator generator, Matrix x, Vector y, IEnumerable`1 examples, Double trainingPct, Int32 total)
   at numl.Learner.Learn(IEnumerable`1 examples, Double trainingPercentage, Int32 repeat, IGenerator generator)
   at MyClass.TestMethod()

**Comment by sethjuarez** [comment]
It looks like the descriptor object is missing the original features. This likely means that it wasn't deserialized properly. Would you mind verifying that the descriptor has the proper features (maybe set a breakpoint and check the contents)? Thanks again!

**Comment by SujataZ** [comment]
Yes, checked it, but could not spot anything wrong with the descriptor. Moreover, it works on the same data, same descriptor for 7 out of 10 times.

One more thing I noticed, this most of the times happens when I set repeatition of learning to more than 1.

Can you please help me to know how can I get the 0.9.15 version project opened in VS 2015? I cant convert my existing applications to new project right now, but wish to debug it with NUML code for this partocular issue.

**Comment by sethjuarez** [comment]
I guess the easiest way right now is to fork the dev branch then create a new blank solution/project and add all of the cs files (it is what I would do). I unfortunately don't have VS15 installed :(

**Comment by SujataZ** [comment]
Ok thanks, will give it a try...

**Comment by sethjuarez** [comment]
Sorry about that! It seems like there is a weird bug in there somewhere....

**Comment by SujataZ** [comment]
It's Ok Seth,

I referred the code from this 

https://github.com/sethjuarez/numl/tree/3414fb433ccf268c183abca78c0e06fa78c607fe

Not sure if below is the fix for it, but I tried with this -

The exception is from this line no. 117 in Descriptor.cs

_return.q.First()_

q is empty in some cases, hence gives exception.

Changed it to _q.FirstOrDefault();_

and then, in _DecisionTreeGenerator.cs_, added a **null check for "property"** after line no. 256

Now, the exception has gone, however the accuracy is always zero :(
struggling over the accuracy issue now further...

**Comment by sethjuarez** [comment]
The real problem is that there is an empty Feature array. That should *never* be empty when going through the learning process. It effectively says there is no data (or columns) to learn from. Is the Feature array empty? If so, we need to find out why that is - can you send me a repro? I want to get to the bottom of the problem.

**Comment by SujataZ** [comment]
Featured array contains 11 elements, with 1 string & 1 datetime field, rest of the fields are double.

Let me try to construct a sample repro for you to refer as the original code cannot be shared.

**Comment by SujataZ** [comment]
Attached is a netcore webapi, please check the "Training" 
[Solution1.zip](https://github.com/sethjuarez/numl/files/874518/Solution1.zip)
 method in ValuesController

Also added a .txt file which contains json to pass to the request as input.

**Comment by sethjuarez** [comment]
I created a more concise project for you that can be run in Visual Studio Code. Some of your Features are highly correlated to the Label. I also noticed that your computed Label is just multiplied by a scalar (which I think is unnecessary.

[SujataTest.zip](https://github.com/sethjuarez/numl/files/876777/SujataTest.zip)

        Accuracy: 1
        No# of Predictions:     6
        ---------------------------------
        Accuracy:               1
        MAE:                    0
        SSE:                    0
        MSE:                    0
        RMSE:                   0
        
        Coef RMSE:              0
        Norm RMSE:              0

        AUC:                    1
        ---------------------------------
        True Positives: 1
        False Positives:        0
        True Negatives: 5
        False Negatives:        0
        ---------------------------------
        Precision:              1
        Recall:         1
        Specificity:            1
        Fallout:                0
        F-Score:                1
        Cross-Entropy:          0


        [TimeTaken, 0.2985]
         |- 4326 <= x < 12105.5
         |       +(0, 0)
         |- 12105.5 <= x < 19885.01
         |      [Field3, 0.3117]
         |       |- 1782 <= x < 1811
         |       |      [Field2, 0.4325]
         |       |       |- 1729.8 <= x < 1744.85
         |       |       |      [Full, 0.5000]
         |       |       |       |- 0 <= x < 0.5
         |       |       |       |       +(1, 1)
         |       |       |       |- 0.5 <= x < 1.01
         |       |       |       |      [Iterations, 0.2740]
         |       |       |       |       |- 2 <= x < 5.5
         |       |       |       |       |       +(0, 0)
         |       |       |       |       |- 5.5 <= x < 9.01
         |       |       |       |       |       +(1, 1)
         |       |       |- 1744.85 <= x < 1759.91
         |       |       |       +(0, 0)
         |       |- 1811 <= x < 1840.01
         |       |       +(0, 0)

It looks like the way you were declaring features was causing some issues. The idea of the project is to separate the learning and the prediction process into two distinct steps. Hope this helps!

**Comment by SujataZ** [comment]
Thanks a ton Seth.

**Comment by sethjuarez** [comment]
My pleasure! Let me know how it all turns out!
